### PR TITLE
Add boolean operators to Mathematical Operators

### PR DESCRIPTION
**Pull request by cmcaine**
A student on exercism struggled to find the boolean not operator for Julia. And perhaps this wouldn't help, but maybe it's worth a go.

### CODE REVIEW

**Comment by cmcaine** [comment]
`a <= b` is equivalent to the material implication operator, but I thought that would be best left out ;)

File: doc/src/manual/mathematical-operations.md [hunk 1]
```diff
@@ -63,6 +63,22 @@ julia> false * Inf
...
+## Boolean Operators
+
+The following [boolean operators](https://en.wikipedia.org/wiki/Boolean_algebra#Operations) are supported on Bool:
+
+| Expression | Name                                  |
+|:---------- |:--------------------------------------|
+| `!x`       | not                                   |
+| `x & y`    | and                                   |
+| `x \| y`   | or                                    |
+| `x ⊻ y`    | xor (exclusive or)                    |
```
Comment (mbauman) [inline doc/src/manual/mathematical-operations.md:75]:
```suggestion
```

These are documented below and I'd rather push folks towards using `&&` and `||` for boolean logic as it has the precedence they likely want.

File: doc/src/manual/mathematical-operations.md [hunk 2]
```diff
@@ -63,6 +63,22 @@ julia> false * Inf
...
+| `x && y`   | [short-circuiting and][short-circuit] |
+| `x \|\| y` | [short-circuiting or][short-circuit]  |
+| `x == y`   | equivalence                           |
+
+Note that Bool is an integer type and all the usual promotions rules and numeric operators are also defined on it (in fact, `&`, `|`, and `⊻` *are* bitwise operators).
```
Comment (cmcaine) [inline doc/src/manual/mathematical-operations.md:80]:
```suggestion
Note that Bool is an integer type and all the usual promotion rules and numeric operators are also defined on it.
```

File: doc/src/manual/mathematical-operations.md (hunk 1 above)
Comment (cmcaine) [reply doc/src/manual/mathematical-operations.md]: Sure. Done.

File: doc/src/manual/mathematical-operations.md [hunk 3]
```diff
@@ -63,6 +63,19 @@ julia> false * Inf
...
+## Boolean Operators
+
+The following [boolean operators](https://en.wikipedia.org/wiki/Boolean_algebra#Operations) are supported on Bool:
+
+| Expression | Name                                  |
+|:---------- |:--------------------------------------|
+| `!x`       | not                                   |
+| `x && y`   | [short-circuiting and][short-circuit] |
+| `x \|\| y` | [short-circuiting or][short-circuit]  |
+| `x == y`   | equivalence                           |
```
Comment (cmcaine) [inline doc/src/manual/mathematical-operations.md:75]:
```suggestion
```

File: doc/src/manual/mathematical-operations.md [hunk 4]
```diff
@@ -63,6 +63,18 @@ julia> false * Inf
...
+The following [boolean operators](https://en.wikipedia.org/wiki/Boolean_algebra#Operations) are supported on Bool:
+
+| Expression | Name                                  |
+|:---------- |:--------------------------------------|
+| `!x`       | not                                   |
+| `x && y`   | [short-circuiting and][short-circuit] |
+| `x \|\| y` | [short-circuiting or][short-circuit]  |
+
+Note that Bool is an integer type and all the usual promotion rules and numeric operators are also defined on it.
+
```
Comment (cmcaine) [inline doc/src/manual/mathematical-operations.md:77]:
```suggestion

[short-circuit]: https://docs.julialang.org/en/v1/manual/control-flow/#Short-Circuit-Evaluation-1
```

Somehow lost this link

**Comment by kimikage** [comment]
I think `||` and `&&` are not operators *in terms of Julia*. I agree with mentioning `&&` and `||` there, but they should clearly indicate that they are not really "operators".

**Comment by cmcaine** [comment]
I feel like that's emphasised already by linking to the short-circuit documentation? Maybe you can come up with a better phrasing?

The overall alternatives are:

1. Status quo: don't explicitly document the Boolean operators
2. Do as I did originally and duplicate information that occurs elsewhere
3. Only show the operators people actually use in if statements and stuff

The PR's current state is at 3. I think any of these are okay (though I would rank them 2 > 3 > 1) and I'm happy to amend it as suggested if someone with merge authority wants to make a judgement call.

**Comment by kimikage** [comment]
> I feel like that's emphasised already by linking to the short-circuit documentation?

The current state of this PR makes me feel like `&&` and `||` are operators in the "same" category as `!`.
That page implicitly requires general programming knowledge. Therefore I think it is important to clarify what's the same and what's different from other programming languages.
(I'm not good at elegant English writing, but what I want to say is that `&&` and `||` can be used as "so-called" binary operators, but strictly speaking, they are control flow keywords, not defined as methods.)

I also think any of those three are OK, as long as there is no misleading about Julia's syntax.

File: doc/src/manual/mathematical-operations.md [hunk 5]
```diff
@@ -63,6 +63,19 @@ julia> false * Inf
...
+The following [boolean operators](https://en.wikipedia.org/wiki/Boolean_algebra#Operations) are supported on Bool:
+
+| Expression | Name                                  |
+|:---------- |:--------------------------------------|
+| `!x`       | not                                   |
+| `x && y`   | [short-circuiting and][short-circuit] |
+| `x \|\| y` | [short-circuiting or][short-circuit]  |
+
+Note that Bool is an integer type and all the usual promotion rules and numeric operators are also defined on it.
```
Comment (dkarrasch) [inline doc/src/manual/mathematical-operations.md:76]:
```suggestion
Note that `Bool` is an integer type and all the usual promotion rules and numeric operators are also defined on it.
```

File: doc/src/manual/mathematical-operations.md [hunk 6]
```diff
@@ -63,6 +63,19 @@ julia> false * Inf
 
 This is useful for preventing the propagation of `NaN` values in quantities that are known to be zero. See [Knuth (1992)](https://arxiv.org/abs/math/9205211) for motivation.
 
+## Boolean Operators
+
+The following [boolean operators](https://en.wikipedia.org/wiki/Boolean_algebra#Operations) are supported on Bool:
```
Comment (dkarrasch) [inline doc/src/manual/mathematical-operations.md:68]:
```suggestion
The following [Boolean operators](https://en.wikipedia.org/wiki/Boolean_algebra#Operations) are supported on `Bool`:
```

**Review by dkarrasch — COMMENTED** [review]
Just a few typographical comments.

**Comment by kimikage** [comment]
If we choose the option 3., it might be reasonable to describe "Boolean Operators" after "Bitwise Operators". By doing so, we can simply declare that "all bitwise operators can be used as boolean operators since `Bool` is a `Integer` type".

Note that the `!x` is in the "Arithmetic Operators" section. I think the `!` should only be in the new "Boolean Operators" section.

File: doc/src/manual/mathematical-operations.md (hunk 6 above)
Comment (cmcaine) [reply doc/src/manual/mathematical-operations.md]: Thanks :)

**Comment by ViralBShah** [comment]
What's next here?

**Comment by JeffBezanson** [comment]
This seems fine to me.

**Pull request merged**